### PR TITLE
[Bug fixes] update input_spec for forward

### PR DESCRIPTION
**Pull request by wj-Mcat**
### PR types

Bug fixes

### PR changes

D2S

### Description

用反射来自动判断是否需要添加 position_ids 的 spec

### CODE REVIEW

File: paddlenlp/transformers/opt/modeling.py [hunk 1]
```diff
@@ -81,7 +81,7 @@ def _expand_mask(mask, tgt_length):
     tgt_length = tgt_length if tgt_length is not None else src_length
 
     expanded_mask = ~(paddle.cast(mask[:, None, None, :], "bool"))
-    expanded_mask = paddle.cast(expanded_mask, dtype=paddle.float32)
+    expanded_mask = paddle.cast(expanded_mask, dtype=paddle.get_default_dtype())
```
Comment (wj-Mcat) [inline paddlenlp/transformers/opt/modeling.py:84]: 最近在做类型提升的工作：
* 以前：fp16 + fp32 = fp16
* 现在：fp16 + fp32 = fp32

所以在模型计算的过程中会出现dtype 不匹配的问题

File: paddlenlp/generation/utils.py [hunk 2]
```diff
@@ -1203,10 +1204,13 @@ def sample(
         return input_ids[:, origin_len:], scores
 
     def _get_model_inputs_spec(self, dtype: str):
-        return {
+        spec = {
             "input_ids": paddle.static.InputSpec(shape=[None, None], dtype="int64"),
             "attention_mask": paddle.static.InputSpec(shape=[None, None], dtype="int64"),
         }
+        if "position_ids" in inspect.getfullargspec(self.forward).args:
```
Comment (wj-Mcat) [inline paddlenlp/generation/utils.py:1211]: 通过反射来判断是否要添加input_ids 这个字段，从而覆盖绝大部分的模型。

**Comment by wj-Mcat** [comment]
## 问题描述

如果在forward PaddleNLP 中的模型时，没有设置 default dtype，则很可能会出现 dtype 类型不匹配的问题，例如 PaddleMix 中的 Blip2 模型在执行 language_model.generate 的时候就会出现此类问题。


## 解决方案

给 paddlenlp 中的模型添加 dtype_guard 即可，示例代码如下所示：

* blip2 的代码：https://github.com/PaddlePaddle/PaddleMIX/blob/develop/paddlemix/models/blip2/modeling.py#L545

```python

from contextlib import contextmanager
@contextmanager
def dtype_guard(dtype="float32"):
    origin_dtype = paddle.get_default_dtype()
    paddle.set_default_dtype(dtype)
    try:
        yield
    finally:
        paddle.set_default_dtype(origin_dtype)

with dtype_guard(self.language_model._dtype):
    outputs = self.language_model.generate(
        inputs_embeds=inputs_embeds,
        attention_mask=attention_mask,
        do_sample=False,
        top_p=0.9,
        decode_strategy="greedy_search",
        temperature=1,
        num_beams=5,
        max_length=30,
        min_length=8,
        eos_token_id=50118,
        repetition_penalty=1,
        length_penalty=1,
        num_return_sequences=1,
    )

return outputs
```

File: paddlenlp/generation/utils.py (hunk 2 above)
Comment (wawltor) [reply paddlenlp/generation/utils.py:1211]: position_ids 在 args里面是None的话，还需要传入position ids 吗？

Comment (wj-Mcat) [reply paddlenlp/generation/utils.py:1211]: 这里会传入position_ids，考虑是：

1. 模型本身是有 position_ids 的属性，需要参与计算。
2. 可以为 None，可是在生成的过程中由于是左 padding，所以 position_id 也是左 padding，如果静态图中默认设置为 None，虽然会重新生成，可是精度会出现问题。
3. position_ids 在大部分的 tokenizer 中会默认返回，在静态图推理阶段会默认将模型对应 tokenizer 的输出都塞入到静态图中去，如果根据 position_ids 为 None 来删除 input_spec，predictor 推理将会直接 down 掉：tokenizer 有 position_id，可是静态图没有 position_ids 的字段。

综合考虑，如果有 position_ids 就直接加入进去。

**Review by wawltor — APPROVED** [review]
LGTM
